lib/src/layer/polygon_layer.dart

```diff
@@ -58,133 +72,190 @@ class PolygonLayer extends StatelessWidget {
-          ..strokeJoin = polygonOpt.strokeJoin;
-
-        _paintLine(canvas, polygonOpt.offsets, borderPaint);
-
-        if (!polygonOpt.disableHolesBorder &&
-            null != polygonOpt.holeOffsetsList) {
-          for (final offsets in polygonOpt.holeOffsetsList!) {
-            _paintLine(canvas, offsets, borderPaint);
-          }
+        paint = Paint()
+          ..style = PaintingStyle.fill
+          ..color = polygon.color;
+
+        // Ideally we'd use `Path.combine(PathOperation.difference, ...)`
+        // instead of evenOdd fill-type, however it creates visual artifacts
+        // using the web renderer.
+        path.fillType = PathFillType.evenOdd;
+
+        path.addPolygon(offsets, true);
+        for (final holeOffsets in holeOffsetsList) {
+          path.addPolygon(holeOffsets, true);
         }
       }
+
+      // Only draw the  border explicitly if it isn't alrady a stroke-style
+      // polygon.
+      if (polygon.borderStrokeWidth > 0.0) {
+        borderPaint = _getBorderPaint(polygon);
+        _paintBorder(borderPath, polygon, offsets, holeOffsetsList);
+      }
+
+      if (polygon.label != null) {
+        // Labels are expensive they mess with draw batching.
+        drawPaths();
+
+        Label.paintText(
+          canvas,
+          offsets,
+          polygon.label,
+          polygon.labelStyle,
+          map.rotationRad,
+          rotate: polygon.rotateLabel,
+          labelPlacement: polygon.labelPlacement,
+        );
+      }
     }
+
+    drawPaths();
   }
 
-  void _paintDottedLine(Canvas canvas, List<Offset> offsets, double radius,
-      double stepLength, Paint paint) {
+  Paint _getBorderPaint(Polygon polygon) {
+    final isDotted = polygon.isDotted;
+    return Paint()
+      ..color = polygon.borderColor
+      ..strokeWidth = polygon.borderStrokeWidth
+      ..strokeCap = polygon.strokeCap
+      ..strokeJoin = polygon.strokeJoin
+      ..style = isDotted ? PaintingStyle.fill : PaintingStyle.stroke;
+  }
+
+  void _paintBorder(ui.Path path, Polygon polygon, List<Offset> offsets,
+      List<List<Offset>> holeOffsetsList) {
+    if (polygon.isDotted) {
+      final borderRadius = (polygon.borderStrokeWidth / 2);
+      final spacing = polygon.borderStrokeWidth * 1.5;
+
+      _paintDottedLine(path, offsets, borderRadius, spacing);
+
+      if (!polygon.disableHolesBorder) {
+        for (final offsets in holeOffsetsList) {
+          _paintDottedLine(path, offsets, borderRadius, spacing);
+        }
+      }
+    } else {
+      _paintLine(path, offsets);
+
+      if (!polygon.disableHolesBorder) {
+        for (final offsets in holeOffsetsList) {
+          _paintLine(path, offsets);
+        }
+      }
+    }
+  }
+
+  void _paintDottedLine(
+      ui.Path path, List<Offset> offsets, double radius, double stepLength) {
     var startDistance = 0.0;
     for (var i = 0; i < offsets.length; i++) {
       final o0 = offsets[i % offsets.length];
```

Comment:
**ACTUALLY FOR LINE 263:**

\-      final totalDistance = _dist(o0, o1);
\+      final totalDistance = (o0 - o1).distance;

There is a built in method that will calculate the distance, this should be more performant and also have cleaner code (since we can remove the helper functions)

Reply:
Happy to rely on library functionality. The performance argument seems a bit far fetched. it's implemented as double get distance => math.sqrt(dx * dx + dy * dy) in (sky_engine/lib/ui/geometry.dart).

Reply:
I haven't profiled it here but generally code runs faster when it is reused (especially in low level libraries), because the CPU cache does not miss code that is run frequently as much. Also generally lower level libraries can have extra compiler optimizations (like using a CPU specific instruction for SQRT for instance) that just "work" correctly where you might only see the performance in some situations. Lastly, sky_engine is going to have more people auditing the performance of the distance function than in flutter_map so any odd tricks or issues will be more likely to be fixed there; let me know if that makes sense or if you have questions!

Reply:
Using standard functionality always makes sense to me. I'm just not buying in the performance argument.

The rest of the argument is more of theoretical nature but who doesn't like a good geeky argument? :) You're arguments are:
 * improved instruction cache hit rate: It's gonna basically be the same
 * improved instructions: any optimization you're thinking of would happen on the math.sqrt level and not on the sky_engine level. Arguable SQRT is slow that's were fast square-root algorithms would come in (quake for the win).
 * better maintainance for sky_engine: sure but in practice there's really only so many ways to implement pythagoras :)

Look i'm happy to use the distance function. I'm just saying that the performance will be the same